Sources/TensorFlow/Layers/Recurrent.swift

```diff
@@ -444,12 +461,99 @@ public struct RecurrentLayer<Cell: RecurrentLayerCell>: Layer {
   }
 }
 
+public protocol Mergeable 
+where Self: Differentiable {
+  @differentiable
+  static func +(lhs: Self, rhs: Self) -> Self
+}
+
+@differentiable
+public func sum<TimeStepOutput: Mergeable>(
+  _ first: TimeStepOutput, 
+  _ second: TimeStepOutput
+) -> TimeStepOutput {
+  return first + second
+}
```

Comment:
Rather than adding a `Mergeable` protocol in this PR, how about conforming `RNNCell.State` types to `AdditiveArithmetic` instead, and using `AdditiveArithmetic.+` instead of a new `func sum`?

Context here: https://github.com/tensorflow/swift-apis/pull/946#discussion_r422509815

My main concern is that a new public protocol `Mergeable` deserves more motivation and review, and it's not quite justified yet in this PR since `sum` is the only merge function. I would prefer if you investigated `Mergeable` and more merge functions in a follow-up PR, to make changes incremental!

Reply:
Oh, I think I see an issue: protocol `Mergeable` is necessary because it requires that `static func +` has a `@differentiable` attribute, unlike `AdditiveArithmetic`.

If that is the case, keeping protocol `Mergeable` in this PR is fine (because it's unavoidable). However, could you try deleting `public func sum` and replacing it with `+` (from `Mergeable`)?

```swift
  public init(_ cell: @autoclosure () -> Cell, merge: @escaping MergeFunction = (+)) {
```

Reply:
Yes, I agree. Sorry for the misunderstanding. I was trying to fix (https://github.com/tensorflow/swift-apis/pull/946/files#diff-85b8b5016877b6312b0c89c0862c5b56R484)

```
  // TODO: Runtime crash when I try to set `sum` function to property
  // @noDerivative public let merge: MergeFunction
```

so I created a new protocol to test if it helps but it didn't. I was also not able to do

```
public init(_ cell: @autoclosure () -> Cell, mergeMode: @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput = +
```

because of 

```
error: expected initial value after '='
```

I will fix it after I figure out tests.
Just wanted to share progress here, so that PR does not look abandoned.

Reply:
Thanks for the context! No worries about "looking abandoned", feel free to make progress at your own pace.

Reply:
Oh, so you figured out the error before I sent a reply. Thanks!

Reply:
```
note: differentiated functions in '@inlinable' functions must be marked '@differentiable' or have a public '@derivative'; this is not possible with a closure, make a top-level function instead
  public init(_ cell: @autoclosure () -> Cell, merge: @escaping MergeFunction = (+))
```

Perhaps I am just doing something wrong, but are you sure  that this should work? 🤔

Reply:
Ah, I understand this error: the expression `(+)` creates a thunk, triggering the error.
In that case, please continue to use the top-level `func sum<T: Mergeable>` function.

I'll discuss this issue with @marcrasi, who added the error, and see if it's still relevant in light of recent differentiation-related changes.